pkg/armhelpers/kubeclient.go

```diff
@@ -0,0 +1,143 @@
+}
+
+//ListPods returns all Pods running on the passed in node
+func (c *KubernetesClientSetClient) ListPods(node *v1.Node) (*v1.PodList, error) {
+	return c.clientset.Core().Pods(metav1.NamespaceAll).List(metav1.ListOptions{
+		FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": node.Name}).String()})
+}
+
+//GetNode returns details about node with passed in name
+func (c *KubernetesClientSetClient) GetNode(name string) (*v1.Node, error) {
+	return c.clientset.Nodes().Get(name, metav1.GetOptions{})
+}
+
+//UpdateNode updates the node in the api server with the passed in info
+func (c *KubernetesClientSetClient) UpdateNode(node *v1.Node) (*v1.Node, error) {
+	return c.clientset.Nodes().Update(node)
+}
+
+//SupportEviction queries the api server to discover if it supports eviction, and returns supported type if it is supported
+func (c *KubernetesClientSetClient) SupportEviction() (string, error) {
+	discoveryClient := c.clientset.Discovery()
+	groupList, err := discoveryClient.ServerGroups()
+	if err != nil {
+		return "", err
+	}
+	foundPolicyGroup := false
+	var policyGroupVersion string
+	for _, group := range groupList.Groups {
+		if group.Name == "policy" {
+			foundPolicyGroup = true
+			policyGroupVersion = group.PreferredVersion.GroupVersion
+			break
+		}
+	}
+	if !foundPolicyGroup {
+		return "", nil
+	}
+	resourceList, err := discoveryClient.ServerResourcesForGroupVersion("v1")
+	if err != nil {
+		return "", err
+	}
+	for _, resource := range resourceList.APIResources {
+		if resource.Name == evictionSubresource && resource.Kind == evictionKind {
+			return policyGroupVersion, nil
+		}
+	}
+	return "", nil
+}
+
+//DeletePod deletes the passed in pod
+func (c *KubernetesClientSetClient) DeletePod(pod *v1.Pod) error {
+	return c.clientset.Core().Pods(pod.Namespace).Delete(pod.Name, &metav1.DeleteOptions{})
+}
+
+//EvictPod evicts the passed in pod using the passed in api version
+func (c *KubernetesClientSetClient) EvictPod(pod *v1.Pod, policyGroupVersion string) error {
+	eviction := &policy.Eviction{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: policyGroupVersion,
+			Kind:       evictionKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      pod.Name,
+			Namespace: pod.Namespace,
+		},
+	}
+	return c.clientset.Policy().Evictions(eviction.Namespace).Evict(eviction)
+}
+
+//GetPod returns the pod
+func (c *KubernetesClientSetClient) getPod(namespace, name string) (*v1.Pod, error) {
+	return c.clientset.Core().Pods(namespace).Get(name, metav1.GetOptions{})
+}
+
+//WaitForDelete waits until all pods are deleted. Returns all pods not deleted and an error on failure
+func (c *KubernetesClientSetClient) WaitForDelete(logger *log.Entry, pods []v1.Pod, usingEviction bool) ([]v1.Pod, error) {
+	var verbStr string
+	if usingEviction {
+		verbStr = "evicted"
+	} else {
+		verbStr = "deleted"
+	}
+	err := wait.PollImmediate(c.interval, c.timeout, func() (bool, error) {
```

Comment:
@JackQuincy will this fail the operation if wait for delete times out (the comment above mentions so)? I don't expect scale to fail with pod deletion fails. This should be best effort IMO.